### ISSUE TITLE
Feature: autosave article

### ISSUE THREAD

**Issue by mateicarpen** [anchorcms/anchor-cms#879]
This is a mandatory feature. I just spent 30 minutes writing an article and by computer blocked and i had to restart. Obviously, there is no article now.

**Comment by TheBrenny** [comment]
That. That is very smart. I actually haven't thought of that! :')

I can see one main problem with it though: to save the article requires a POST - I'm not sure if we can post without requiring a redirect on the page...

But nevertheless, I will look into this! :D

**Comment by mateicarpen** [comment]
Hey.

You can do an ajax post request that would just update the db, and not the view (so no reload necessary).

Alternatively, you could just autosave fields in a cookie, and populate the form from that, if they leave the page by mistake.

Hope this helps.

**Comment by TheBrenny** [comment]
I have no history in ajax or how cookies save and open.. Do you reckon you could make this one up and benefit the world? :D

**Comment by Cam** [comment]
I think you could probably store the info in localStorage with JSON fairly easily, along with the pages' unique identifier. It would be quite fitting, with the unique simplicity of Anchor, to have a live autosaving feature built in.

**Comment by TheBrenny** [comment]
I think AJAX would be better suited for this sort of thing. My reasoning is that it updates the database.

@Cam, here's where the two methods differ. Say you're on your desktop PC at home, and start an article. You don't finish it because you need to find a picture or something at work. Instead of uploading the picture to a file storage service (USB or Cloud-Based) and then waiting until you get home, you can quickly finish it off right there because everything is saved server side.

I'm reading W3's AJAX tutorial, so seeing it as JavaScript and an XMLHttpRequest puts me in a more hopeful mood of being able to complete this sort of thing. But if @mateicarpen has already started this feature, I'll gladly step back and let him finish it. I have a bug that I'm trying to nail out anyway.

**Comment by Cam** [comment]
@TheBrenny definitely... but I think you either want to go fully one way or the other.

Either the live-saved data is going straight into the DB, or it isn't. Anything else is really just a session saving device. Both cookies and localStorage have issues otherwise.

Either way, AJAX is definitely the right way to go.

**Comment by mateicarpen** [comment]
Yes, i can look at this, it's a fairly easy task. I should be able to push something in a few days.

**Comment by TheBrenny** [comment]
Cheers! If you can, also implement this feature for page authoring as well.

**Comment by CraigChilds94** [comment]
I personally feel like you should just add something to this file: https://github.com/anchorcms/anchor-cms/blob/0.9-dev/anchor/views/assets/js/editor.js which calls `.submit()` on the editor form, say, every 30 seconds. Then displays a notification reading "Post autosaved".

**Comment by Cam** [comment]
Yeah @CraigChilds94. Definitely. That's far more logical, as it goes straight into the DB without any fussing around. I'm not sure what the purpose of temporary storage was really. I guess it was just a red herring.

**Comment by TimohStudios** [comment]
Sounds like a great feature, that isn't too hard to implement. Don't forget to always autosave as a draft!

**Comment by TheBrenny** [comment]
I'm not 100% certain, but I think the submit button is asynchronous anyway?

I know that in Chrome it won't reload the page, and I think that it could be because it will redirect you to the exact same page...

I've not tested this in other browsers, and perhaps this could be a chrome specific feature? It caches previously visited pages, so logging out and pressing the back button shows the previous page rather than the login screen - however this could be a CSRF issue...

Things to investigate... :3

**Comment by IwfY** [comment]
@TimohStudios your comment made me think about an interesting issue. If you edit an article the autosave would need to save to a new (temporary) article. Otherwise a previously released article wouldn't be available anymore.

Consider you are editing an article and halfway through you decide that you don't want to submit your changes. How do you offer this? And you would have to communicate to the author that leaving the page will leave your article changed but inaccessible.

**Comment by TheBrenny** [comment]
@IwfY, I feel that if you are making changes and then suddenly think, "nah, I'll actually pass.", you would end up mashing the `CTRL` + `Z` buttons. :stuck_out_tongue_winking_eye:
